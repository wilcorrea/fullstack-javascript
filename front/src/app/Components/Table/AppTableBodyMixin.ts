import { CreateElement, VNode } from 'vue'
import { TableField, Record } from 'src/app/Agnostic/interfaces'
import { block } from './index'

/**
 */
export default {
  /**
   */
  methods: {
    /**
     * @param {CreateElement} h
     * @param {Array<TableField>} columns
     * @param records
     * @returns {VNode}
     */
    tbody (h: CreateElement, columns: Array<TableField>, records: any): VNode {
      if (!records) {
        return h('tbody', this.tbodyEmpty(h))
      }
      // <transition-group tag="ul" name="list"
      const data = { props: { tag: 'tbody', name: 'app-table__list' } }
      const children = this.tbodyRows(h, columns, records)
      return h('transition-group', data, children)
    },
    /**
     * @param {CreateElement} h
     * @returns {VNode}
     */
    tbodyEmpty (h: CreateElement): Array<VNode> {
      return [h('tr', [h('td', 'no-data')])]
    },
    /**
     * @param {CreateElement} h
     * @param {Array<TableField>} columns
     * @param {Array<Record>} records
     * @returns {Array<VNode>}
     */
    tbodyRows (h: CreateElement, columns: Array<TableField>, records: Array<Record>): Array<VNode> {
      return records.map((record: Record, index: number) => this.tbodyRow(h, columns, record, index))
    },
    /**
     * @param {CreateElement} h
     * @param {Array<TableField>} columns
     * @param {Record} record
     * @param {number} index
     * @returns {VNode}
     */
    tbodyRow (h: CreateElement, columns: Array<TableField>, record: Record, index: number) {
      const data = { key: index }
      const children = columns.map((column: TableField) => {
        const data = { class: `${block}__${column.key}--value` }
        // @ts-ignore
        const style: any = this.parseColumnStyle(column, 'td')
        // @ts-ignore
        data.style = { style: { width: style['max-width'] } }
        return h('td', data, this.tbodyRowValue(h, style, record, column))
      })
      return h('tr', data, children)
    },
    /**
     * @param {CreateElement} h
     * @param style
     * @param {Record} record
     * @param {TableField} column
     * @returns {Array<VNode>}
     */
    tbodyRowValue (h: CreateElement, style: any, record: Record, column: TableField): Array<VNode> {
      const data = { class: `${block}__td`, style }
      // @ts-ignore
      return [h('div', data, this.tbodyRowValueCell(record, column, h))]
    }
  }
}
