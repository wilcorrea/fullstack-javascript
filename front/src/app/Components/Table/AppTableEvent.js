/**
 */
export default {
  /**
   */
  mounted () {
    if (!this.pressHold || !(this.$el && typeof this.$el.querySelector === 'function')) {
      return
    }
    let element = this.$el.querySelector('[move]')
    if (!element) {
      return
    }
    this.element = element

    let x = 0
    let y = 0
    let top = 0
    let left = 0

    let draggingFunction = (e) => {
      element.scrollLeft = left - e.pageX + x
      element.scrollTop = top - e.pageY + y
    }

    this.dragging = (event) => {
      this.element.style['cursor'] = 'move'
      this.element.style['user-select'] = 'none'
      event.preventDefault()
      y = event.pageY
      x = event.pageX
      top = element.scrollTop
      left = element.scrollLeft

      document.addEventListener('mousemove', draggingFunction)

      this.stopDragging = () => {
        document.removeEventListener('mousemove', draggingFunction)
        this.element.style['cursor'] = 'inherit'
        this.element.style['user-select'] = 'inherit'
      }
      document.addEventListener('mouseup', this.stopDragging)
      element.addEventListener('mouseleave', this.stopDragging)
    }

    let timerID
    let counter = 0
    const pressingDown = (e) => {
      this.event = e
      requestAnimationFrame(timer)
    }
    const notPressingDown = (e) => {
      cancelAnimationFrame(timerID)
      counter = 0
    }
    const timer = () => {
      if (counter < 50) {
        timerID = requestAnimationFrame(timer)
        counter++
        return
      }
      this.dragging(this.event)
    }

    element.addEventListener('mousedown', pressingDown, false)
    element.addEventListener('mouseup', notPressingDown, false)
    element.addEventListener('mouseleave', notPressingDown, false)
  },
  /**
   */
  destroyed () {
    if (!this.element) {
      return
    }
    this.element.removeEventListener('pressHoldEvent', this.dragging)
    this.element.removeEventListener('mouseleave', this.stopDragging)
    if (this.stopDragging) {
      document.removeEventListener('mouseup', this.stopDragging)
    }
  }
}
