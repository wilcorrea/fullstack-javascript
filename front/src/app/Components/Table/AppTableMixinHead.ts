import { CreateElement, VNode } from 'vue'
import { TableField } from 'src/app/Agnostic/interfaces'
import { block } from './index'

/**
 */
export default {
  /**
   */
  methods: {
    /**
     * @param {CreateElement} h
     * @param {Array<TableField>} columns
     * @returns {VNode}
     */
    thead (h: CreateElement, columns: Array<TableField>): VNode {
      const children = columns.map((column: TableField) => {
        // @ts-ignore
        const style: any = this.parseColumnStyle(column, 'th')
        const data = {
          class: `${block}__th ${block}__${column.key}`,
          style: {
            width: style['max-width'],
            textAlign: style.textAlign ? style.textAlign : style['text-align']
          }
        }
        return h('th', { style: { width: style['max-width'] } }, [h('div', data, column.label)])
      })
      return h('thead', [h('tr', children)])
    }
  }
}
