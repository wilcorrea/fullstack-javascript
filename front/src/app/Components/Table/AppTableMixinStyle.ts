import { TableField } from 'src/app/Agnostic/interfaces'
import { get } from 'src/app/Util/general'

/**
 */
export default {
  /**
   */
  methods: {
    /**
     * @param {TableField} column
     * @param {string} type
     * @returns {any}
     */
    parseColumnStyle (column: TableField, type: string): any {
      if (typeof column.stylish === 'function') {
        return column.stylish(type)
      }

      const map = {
        width: 'max-width',
        height: 'height',
        align: 'text-align'
      }
      const reducer = (accumulator: any, key: any) => {
        const value = get(column, key)
        if (value) {
          // @ts-ignore
          accumulator[map[key]] = value
        }
        return accumulator
      }
      return Object.keys(map).reduce(reducer, {})
    }
  }
}
