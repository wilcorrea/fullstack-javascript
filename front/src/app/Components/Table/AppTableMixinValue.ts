import { CreateElement } from 'vue'
import { TableField, Record } from 'src/app/Agnostic/interfaces'
import { get } from 'src/app/Util/general'

/**
 */
export default {
  /**
   */
  methods: {
    /**
     * @param {Record} record
     * @param {TableField} column
     * @param {CreateElement} h
     * @returns {any}
     */
    tbodyRowValueCell (record: Record, column: TableField, h: CreateElement): any {
      const value = this.getValue(column, record)
      const key = column.key
      if (!column.formatter) {
        return value
      }
      return column.formatter({ h, value, record, column, key })
    },
    /**
     * @param {TableField} column
     * @param {Record} record
     * @returns {any}
     * @private
     */
    getValue (column: TableField, record: Record) {
      return get(record, column.key, {})
    }
  }
}
