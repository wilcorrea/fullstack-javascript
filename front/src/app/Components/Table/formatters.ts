import { CreateElement, VNode } from 'vue'
import { TableField } from 'src/app/Agnostic/interfaces'
import { browse } from 'src/app/Util/general'
import { block } from './index'

/**
 * @param {CreateElement} h
 * @param {TableField} column
 * @param value
 * @param {string} path
 * @returns {VNode[]}
 */
export const link = (h: CreateElement, column: TableField, value: any, path: string) => {
  const key = column.key
  // @ts-ignore
  const domProps = { href: path }
  const on = {
    click: ($event: Event) => {
      $event.stopPropagation()
      $event.preventDefault()
      browse(path)
    }
  }
  return [h('a', { class: `${block}__a`, domProps, on }, value)]
}

/**
 * @param {Object} args
 * @returns {Array<VNode>}
 */
// @ts-ignore
export const html = ({ h, column, record, value, key, render }): Array<VNode> => {
  return [h('span', { domProps: { innerHTML: render({ h, column, record, value, key, render }) } })]
}
