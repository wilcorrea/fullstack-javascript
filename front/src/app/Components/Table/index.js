/**
 * @type {string}
 */
export const block = 'app-table'
