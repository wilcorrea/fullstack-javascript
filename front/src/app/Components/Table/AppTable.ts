import { CreateElement, VNode } from 'vue'
import AppTableBody from 'src/app/Components/Table/AppTableBodyMixin'
import AppTableHead from 'src/app/Components/Table/AppTableMixinHead'
import AppTableValue from 'src/app/Components/Table/AppTableMixinValue'
import AppTableStyle from 'src/app/Components/Table/AppTableMixinStyle'
import AppTableEvent from './AppTableEvent'
import { block } from './index'

/**
 */
export default {
  /**
   */
  name: 'AppTable',
  /**
   */
  mixins: [
    AppTableHead, AppTableBody, AppTableValue, AppTableStyle, AppTableEvent
  ],
  /**
   */
  props: {
    columns: {
      type: Array,
      required: true
    },
    records: {
      type: Array,
      default: () => []
    },
    loading: {
      type: Boolean,
      default: false
    },
    loader: {
      type: String,
      default: '<div class="ld ld-ring ld-vortex-in"></div>'
    },
    pressHold: {
      type: Boolean,
      default: true
    }
  },
  /**
   */
  methods: {
    /**
     * @param {CreateElement} h
     * @returns {VNode | undefined}
     */
    table (h: CreateElement): VNode | undefined {
      // @ts-ignore
      if (!this.columns) {
        return
      }
      const data = { class: block }
      const children = [
        this.tableContainer(h),
        this.tableShadow(h)
      ]

      return h('div', data, children)
    },
    /**
     * @param {CreateElement} h
     * @returns {VNode}
     */
    tableContainer (h: CreateElement): VNode {
      const root = `${block}__container`
      // @ts-ignore
      const loading = this.loading ? `${root}--loading` : ''
      const data = { class: [root, loading], attrs: { move: true } }
      // @ts-ignore
      const table = [this.thead(h, this.columns), this.tbody(h, this.columns, this.records)]
      const children = [
        h('table', table),
        this.tableLoading(h)
      ]
      return h('div', data, children)
    },
    /**
     * @param {CreateElement} h
     * @returns {VNode}
     */
    tableLoading (h: CreateElement): VNode {
      // @ts-ignore
      const visible = this.loading ? `${block}__loading--visible` : ''
      // @ts-ignore
      const html = `<div class="${block}__loading-container">${this.loader}</div>`
      const data = {
        class: [`${block}__loading`, visible],
        domProps: { innerHTML: html }
      }
      return h('div', data)
    },
    /**
     * @param {CreateElement} h
     * @returns {VNode}
     */
    tableShadow (h: CreateElement): VNode {
      // @ts-ignore
      const loading = this.loading ? `${block}__shadow--loading` : ''
      return h('div', { class: [`${block}__shadow`, loading] })
    }
  },
  /**
   * @param {CreateElement} h
   * @returns {VNode}
   */
  render (h: CreateElement): VNode {
    // @ts-ignore
    return this.table(h)
  }
}
