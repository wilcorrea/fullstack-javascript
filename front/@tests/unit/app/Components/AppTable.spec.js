import { shallowMount } from '@vue/test-utils'
import AppTable from 'src/app/Components/Table/AppTable'
import { html, link } from 'src/app/Components/Table/formatters'

// import $lang from 'src/lang'

describe('AppTable', () => {
  const spy = jest.spyOn(console, 'error')
  shallowMount(AppTable)

  /**
   * @test
   */
  it('"columns" is required', () => {
    expect(spy).toBeCalledWith(expect.stringContaining('[Vue warn]: Missing required prop: "columns"'))
  })

  const path = 'root'
  const fieldId = { key: 'id', label: 'ID', formatter: ({ value }) => value }
  const tabletLink = {
    key: 'name',
    label: 'tableLink',
    formatter: ({ h, column, value }) => link(h, column, value, `${path}/${value}`)
  }
  const tableFormat = {
    key: 'phone',
    label: 'tableFormat',
    formatter: ({ value, record, key }) => `${value}/${key}/${record[key]}`
  }
  const tableHtml = {
    key: 'address',
    label: 'tableHtml',
    formatter: ({ h, column, value }) => {
      // noinspection JSCheckFunctionSignatures
      return html({ h, value, render: ({ value }) => `<b>${value}</b>` })
    }
  }
  const tableComponent = {
    key: 'status',
    label: 'tableHtml',
    formatter: ({ h, value }) => [h('b', value)]
  }
  const columns = [fieldId, tabletLink, tableFormat, tableHtml, tableComponent]
  const record = { id: 1, name: 'Abc', phone: '333', address: '= )', status: 'face' }
  const records = [record]
  const wrapper = shallowMount(AppTable, { propsData: { columns, records } })

  /**
   * @test
   */
  it('create the table head', () => {
    expect(wrapper.find('table > thead > tr > th > .app-table__th').text())
      .toMatch(fieldId.label)
    expect(wrapper.find(`table > thead > tr > th > .app-table__${fieldId.key}`).text())
      .toMatch(fieldId.label)
  })

  const pathValue = (key, edge = '') => {
    return `tr > td.app-table__${key}--value > .app-table__td${edge}`
  }
  /**
   * @test
   */
  it('create the table body', () => {
    expect(wrapper.find(pathValue(fieldId.key)).text())
      .toMatch(String(record[fieldId.key]))
  })
  it('create the table body: tableLink', () => {
    expect(wrapper.find(pathValue(tabletLink.key, ' > a')).attributes('href'))
      .toMatch(`${path}/${record[tabletLink.key]}`)

    expect(wrapper.vm.$el).toMatchSnapshot('tableLink')
  })
  it('create the table body: tableFormat', () => {
    expect(wrapper.find(pathValue(tableFormat.key)).text())
      .toMatch(`${record[tableFormat.key]}/${tableFormat.key}/${record[tableFormat.key]}`)

    expect(wrapper.vm.$el).toMatchSnapshot('tableFormat')
  })
  it('create the table body: tableHtml', () => {
    expect(wrapper.find(pathValue(tableHtml.key, ' > span')).html())
      .toMatch(`<b>${record[tableHtml.key]}</b>`)

    expect(wrapper.vm.$el).toMatchSnapshot('tableHtml')
  })
  it('create the table body: tableComponent', () => {
    expect(wrapper.find(pathValue(tableComponent.key)).html())
      .toMatch(`<b>${record[tableComponent.key]}</b>`)

    expect(wrapper.vm.$el).toMatchSnapshot('tableComponent')
  })

  /**
   * @test
   */
  test('should render content correctly', () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })
})
