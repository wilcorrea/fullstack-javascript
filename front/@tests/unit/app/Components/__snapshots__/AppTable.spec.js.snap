// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppTable create the table body: tableComponent: tableComponent 1`] = `
<div
  class="app-table"
>
  <div
    class="app-table__container"
    move="true"
  >
    <table>
      <thead>
        <tr>
          <th>
            <div
              class="app-table__th app-table__id"
            >
              ID
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__name"
            >
              tableLink
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__phone"
            >
              tableFormat
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__address"
            >
              tableHtml
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__status"
            >
              tableHtml
            </div>
          </th>
        </tr>
      </thead>
      <span>
        <tr>
          <td
            class="app-table__id--value"
          >
            <div
              class="app-table__td"
            >
              1
            </div>
          </td>
          <td
            class="app-table__name--value"
          >
            <div
              class="app-table__td"
            >
              <a
                class="app-table__a"
                href="root/Abc"
              >
                Abc
              </a>
            </div>
          </td>
          <td
            class="app-table__phone--value"
          >
            <div
              class="app-table__td"
            >
              333/phone/333
            </div>
          </td>
          <td
            class="app-table__address--value"
          >
            <div
              class="app-table__td"
            >
              <span>
                <b>
                  = )
                </b>
              </span>
            </div>
          </td>
          <td
            class="app-table__status--value"
          >
            <div
              class="app-table__td"
            >
              <b>
                face
              </b>
            </div>
          </td>
        </tr>
      </span>
    </table>
    <div
      class="app-table__loading"
    >
      <div
        class="app-table__loading-container"
      >
        <div
          class="ld ld-ring ld-vortex-in"
        />
      </div>
    </div>
  </div>
  <div
    class="app-table__shadow"
  />
</div>
`;

exports[`AppTable create the table body: tableFormat: tableFormat 1`] = `
<div
  class="app-table"
>
  <div
    class="app-table__container"
    move="true"
  >
    <table>
      <thead>
        <tr>
          <th>
            <div
              class="app-table__th app-table__id"
            >
              ID
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__name"
            >
              tableLink
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__phone"
            >
              tableFormat
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__address"
            >
              tableHtml
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__status"
            >
              tableHtml
            </div>
          </th>
        </tr>
      </thead>
      <span>
        <tr>
          <td
            class="app-table__id--value"
          >
            <div
              class="app-table__td"
            >
              1
            </div>
          </td>
          <td
            class="app-table__name--value"
          >
            <div
              class="app-table__td"
            >
              <a
                class="app-table__a"
                href="root/Abc"
              >
                Abc
              </a>
            </div>
          </td>
          <td
            class="app-table__phone--value"
          >
            <div
              class="app-table__td"
            >
              333/phone/333
            </div>
          </td>
          <td
            class="app-table__address--value"
          >
            <div
              class="app-table__td"
            >
              <span>
                <b>
                  = )
                </b>
              </span>
            </div>
          </td>
          <td
            class="app-table__status--value"
          >
            <div
              class="app-table__td"
            >
              <b>
                face
              </b>
            </div>
          </td>
        </tr>
      </span>
    </table>
    <div
      class="app-table__loading"
    >
      <div
        class="app-table__loading-container"
      >
        <div
          class="ld ld-ring ld-vortex-in"
        />
      </div>
    </div>
  </div>
  <div
    class="app-table__shadow"
  />
</div>
`;

exports[`AppTable create the table body: tableHtml: tableHtml 1`] = `
<div
  class="app-table"
>
  <div
    class="app-table__container"
    move="true"
  >
    <table>
      <thead>
        <tr>
          <th>
            <div
              class="app-table__th app-table__id"
            >
              ID
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__name"
            >
              tableLink
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__phone"
            >
              tableFormat
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__address"
            >
              tableHtml
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__status"
            >
              tableHtml
            </div>
          </th>
        </tr>
      </thead>
      <span>
        <tr>
          <td
            class="app-table__id--value"
          >
            <div
              class="app-table__td"
            >
              1
            </div>
          </td>
          <td
            class="app-table__name--value"
          >
            <div
              class="app-table__td"
            >
              <a
                class="app-table__a"
                href="root/Abc"
              >
                Abc
              </a>
            </div>
          </td>
          <td
            class="app-table__phone--value"
          >
            <div
              class="app-table__td"
            >
              333/phone/333
            </div>
          </td>
          <td
            class="app-table__address--value"
          >
            <div
              class="app-table__td"
            >
              <span>
                <b>
                  = )
                </b>
              </span>
            </div>
          </td>
          <td
            class="app-table__status--value"
          >
            <div
              class="app-table__td"
            >
              <b>
                face
              </b>
            </div>
          </td>
        </tr>
      </span>
    </table>
    <div
      class="app-table__loading"
    >
      <div
        class="app-table__loading-container"
      >
        <div
          class="ld ld-ring ld-vortex-in"
        />
      </div>
    </div>
  </div>
  <div
    class="app-table__shadow"
  />
</div>
`;

exports[`AppTable create the table body: tableLink: tableLink 1`] = `
<div
  class="app-table"
>
  <div
    class="app-table__container"
    move="true"
  >
    <table>
      <thead>
        <tr>
          <th>
            <div
              class="app-table__th app-table__id"
            >
              ID
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__name"
            >
              tableLink
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__phone"
            >
              tableFormat
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__address"
            >
              tableHtml
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__status"
            >
              tableHtml
            </div>
          </th>
        </tr>
      </thead>
      <span>
        <tr>
          <td
            class="app-table__id--value"
          >
            <div
              class="app-table__td"
            >
              1
            </div>
          </td>
          <td
            class="app-table__name--value"
          >
            <div
              class="app-table__td"
            >
              <a
                class="app-table__a"
                href="root/Abc"
              >
                Abc
              </a>
            </div>
          </td>
          <td
            class="app-table__phone--value"
          >
            <div
              class="app-table__td"
            >
              333/phone/333
            </div>
          </td>
          <td
            class="app-table__address--value"
          >
            <div
              class="app-table__td"
            >
              <span>
                <b>
                  = )
                </b>
              </span>
            </div>
          </td>
          <td
            class="app-table__status--value"
          >
            <div
              class="app-table__td"
            >
              <b>
                face
              </b>
            </div>
          </td>
        </tr>
      </span>
    </table>
    <div
      class="app-table__loading"
    >
      <div
        class="app-table__loading-container"
      >
        <div
          class="ld ld-ring ld-vortex-in"
        />
      </div>
    </div>
  </div>
  <div
    class="app-table__shadow"
  />
</div>
`;

exports[`AppTable should render content correctly 1`] = `
<div
  class="app-table"
>
  <div
    class="app-table__container"
    move="true"
  >
    <table>
      <thead>
        <tr>
          <th>
            <div
              class="app-table__th app-table__id"
            >
              ID
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__name"
            >
              tableLink
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__phone"
            >
              tableFormat
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__address"
            >
              tableHtml
            </div>
          </th>
          <th>
            <div
              class="app-table__th app-table__status"
            >
              tableHtml
            </div>
          </th>
        </tr>
      </thead>
      <span>
        <tr>
          <td
            class="app-table__id--value"
          >
            <div
              class="app-table__td"
            >
              1
            </div>
          </td>
          <td
            class="app-table__name--value"
          >
            <div
              class="app-table__td"
            >
              <a
                class="app-table__a"
                href="root/Abc"
              >
                Abc
              </a>
            </div>
          </td>
          <td
            class="app-table__phone--value"
          >
            <div
              class="app-table__td"
            >
              333/phone/333
            </div>
          </td>
          <td
            class="app-table__address--value"
          >
            <div
              class="app-table__td"
            >
              <span>
                <b>
                  = )
                </b>
              </span>
            </div>
          </td>
          <td
            class="app-table__status--value"
          >
            <div
              class="app-table__td"
            >
              <b>
                face
              </b>
            </div>
          </td>
        </tr>
      </span>
    </table>
    <div
      class="app-table__loading"
    >
      <div
        class="app-table__loading-container"
      >
        <div
          class="ld ld-ring ld-vortex-in"
        />
      </div>
    </div>
  </div>
  <div
    class="app-table__shadow"
  />
</div>
`;
